FSE: Use get_theme_support instead of theme whitelist

glendaviesnz opened the pull request:
#### Changes proposed in this Pull Request

* Remove the theme whitelist and use get_theme_support( 'full-site-editing' ) to determine theme compatibility

#### Testing instructions

* Apply this patch and https://github.com/Automattic/themes/pull/1351/files to your local FSE dev env
* Check that enabling Modern Business, Maywood and Varia with FSE plugin enabled still populates the template data and shows header and footer template parts when viewing or adding a page

gwwar commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php:
```diff
@@ -99,7 +88,7 @@ public static function get_instance() {
 	 */
 	public function is_supported_theme( $theme_slug ) {
 		$theme_slug = $this->normalize_theme_slug( $theme_slug );
```
We don't need this line

gwwar replied:
Actually I suppose the name function doesn't make sense without this. How about:

```php
//keep around temporarily for migration purposes
public function is_supported_theme( $theme_slug ) {
    return get_theme_support( 'full-site-editing' );
}

//suggested naming
public static function is_full_site_editing_active() {
    if ( apply_filters( 'a8c_disable_full_site_editing', false ) ) { 
        return false;
    }
    return get_theme_support( 'full-site-editing' );
}
```

gwwar replied:
@kwight @mattwiebe let me know if the `apply_filters` bit makes sense

mattwiebe replied:
(Almost not posting since this might just muddy the waters with other concerns, but it does also address the current issue)

I don't think that this should have the same filter name as used in [the loader](https://github.com/Automattic/wp-calypso/blob/5cde6adfda0f6e3754b3a6e223d887365b0b6045/apps/full-site-editing/full-site-editing-plugin/full-site-editing-plugin.php#L36-L38). It mixes concerns (loading on the one hand, marking active on another). We have to be able to load the plugin to even run this function, presently.

Also, since the plugin is already namespaced, I think we could just go with `is_active` for the function name. Something like 

```php
//keep around temporarily for migration purposes
public function is_supported_theme( $theme_slug ) { //in reality is_current_theme_supported
    return current_theme_supports( 'full-site-editing' );
}

//suggested naming
public static function is_active() {
    if ( ! apply_filters( 'a8c_is_full_site_editing_active', true ) ) { 
        return false;
    }
    return current_theme_supports( 'full-site-editing' );
}
// called with:
\A8C\FSE\Full_Site_Editing::is_active();
```

And really, at that point, we probably want to ensure that at least the `\A8C\FSE\Full_Site_Editing` class is always loaded. Because otherwise we have this function that we can't even call. Or maybe we just put a version of it in the base plugin file. Hmm.

gwwar replied:
We can do the rest of the tidying across multiple PRs. For example this one just updates is_supported_theme and maybe adds the better is_active check that no one calls yet. (I'd be okay on a different PR to add that). 

Just noting that naming is completely off atm :D But I think callers should be okay.

gwwar commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php:
```diff
@@ -39,17 +39,6 @@ class Full_Site_Editing {
-		'maywood',
-		'modern-business',
-		'varia',
-	];
```
❤️

glendaviesnz commented:
I have updated the check to use current_theme_supports() and added a deprecation message to the is_supported_theme method. I think it would be better to add the is_active method after the  'a8c_is_full_site_editing_active' filter is available to avoid any confusion.

gwwar commented:
For testing a new site creation, we also need to apply theme changes on the sandbox

gwwar commented:
If we land early, header/footer templates will load forever.

Copons commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php:
```diff
@@ -93,13 +82,16 @@ public static function get_instance() {
-		$theme_slug = $this->normalize_theme_slug( $theme_slug );
-		return in_array( $theme_slug, self::SUPPORTED_THEMES, true );
+	// phpcs:disable
+	public function is_supported_theme( $theme_slug = null ) {
```
Nitpick, but is there a reason why we need to declare the function parameter?

gwwar replied:
For backwards compatibility. Callers still expect this to be checking against a theme whitelist. We can remove the whole thing after we add in cleaner checks with `is_fse_active` and `is_fse_eligible` for a site.

Copons replied:
I mean, `$theme_slug` is never used in the function though.
Would PHP complain if someone called a function without parameters with a parameter?
E.g.
```php
function foo() {
  // ...
}

foo( 'bar' );
```

gwwar replied:
You'd then question why `is_supported_theme()` has no args right? Let's just keep this as is for now, since we plan on removing the entire function

Copons commented:
I've opened https://github.com/Automattic/themes/pull/1384 to remove FSE support from Modern Business.
I've noticed a couple of things, testing Modern Business on that PR together with this:

1) The page editor still shows the template parts, loading forever. @gwwar is this what you meant in [your comment](https://github.com/Automattic/wp-calypso/pull/35856#issuecomment-526399404)?

2) The front end is still adding the `fse-enabled` class to the body and, until I've removed them altogether, tried rendering the FSE template parts (see point 3: apparently the `A8C\FSE\WP_Template` class still exists) instead of the theme's ones.

3) Currently, the front end relies on `if ( class_exists( 'A8C\FSE\WP_Template' ) )` checks (see [Varia](https://github.com/Automattic/themes/blob/eaf9a10037d4e08fdb75d4eabc366a9f01622203/varia/header.php#L26)). Should we eventually change that check with the upcoming `is_active` function?

gwwar commented:
> The page editor still shows the template parts, loading forever. 

I think we're missing a case with only trying to populate on theme switch. We can still mark a site as FSE eligible with them having that specific theme active, or turn on theme support.

> Should we eventually change that check with the upcoming is_active function?

I'm currently overloading `a8c_disable_full_site_editing` for this purpose, since logic does differ between site types. We can refactor further, but for now this should be a no-op

D32356-code
